cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -0,0 +1,311 @@
+ *
+ * @param certificateARN
+ * @param {Set<String>} aliases for the service.
+ */
+async function waitForValidationOptionsToBeReady(certificateARN, aliases) {
+    let expectedCount = aliases.size + 1; // Expect one validation option for each alias and the cert domain.
+
+    let attempt; // TODO: This wait loops could be further abstracted.
+    for (attempt = 0; attempt < ATTEMPTS_VALIDATION_OPTIONS_READY; attempt++) {
+        let readyCount = 0;
+        const { Certificate } = await acm.describeCertificate({
+            CertificateArn: certificateARN,
+        }).promise();
+        const options = Certificate.DomainValidationOptions || [];
+        options.forEach(option => {
+            if (option.ResourceRecord && (aliases.has(option.DomainName) || option.DomainName === certificateDomain)) {
+                readyCount++;
+            }
+        })
+        if (readyCount === expectedCount) {
+            return options;
+        }
+
+        // Exponential backoff with jitter based on 200ms base
+        // component of backoff fixed to ensure minimum total wait time on
+        // slow targets.
+        const base = Math.pow(2, attempt);
+        await sleep(random() * base * 50 + base * 150);
+    }
+    throw new Error(`resource validation records are not ready after ${attempt} tries`);
+}
+
+/**
+ * Validate the certificate and insert the alias records
+ *
+ * @param {Array<Object>} validationOptions
+ * @param {String} certificateARN
+ * @param {String} loadBalancerDNS
+ * @param {String} loadBalancerHostedZone
+ */
+async function activate(validationOptions, certificateARN, loadBalancerDNS, loadBalancerHostedZone) {
+    let promises = [];
+    for (let option of validationOptions) {
+        promises.push(activateOption(option, loadBalancerDNS, loadBalancerHostedZone));
+    }
+    await Promise.all(promises);
+    console.log("finished upserting records");
+
+    await acm.waitFor("certificateValidated", {
+        // Wait up to 9 minutes and 30 seconds
+        $waiter: {
+            delay: DELAY_CERTIFICATE_VALIDATED,
+            maxAttempts: ATTEMPTS_CERTIFICATE_VALIDATED,
+        },
+        CertificateArn: certificateARN,
+    }).promise();
+}
+
+/**
+ * Upsert the validation record for the alias, as well as adding the A record if the alias is not the default certificate domain.
+ *
+ * @param {Object} option
+ * @param {String} loadBalancerDNS
+ * @param {String} loadBalancerHostedZone
+ */
+async function activateOption(option, loadBalancerDNS, loadBalancerHostedZone) {
+    let changes = [{
+        Action: "UPSERT",
+        ResourceRecordSet: {
+            Name: option.ResourceRecord.Name,
+            Type: option.ResourceRecord.Type,
+            TTL: 60,
+            ResourceRecords: [
+                {
+                    Value: option.ResourceRecord.Value,
+                },
+            ],
+        }
+    }];
+
+    if (option.DomainName !== certificateDomain) {
```

Comment:
For me to understand the domain name is indeed updated when any `alias` is set right?

Reply:
Yes the `option.DomainName` would be the alias domain name. 

The `option` is given by `acm` when it provides us with the validation options for each aliases. Each option should contain at least the following information:
1. The domain it's validated (i.e. the alias)
2. The validation record name
3. The validation record value